txpipe/psf_diagnostics.py

```diff
@@ -172,6 +172,415 @@ def plot_histogram(self, function, output_name, xlabel, edges):
+        Xall  = np.concatenate([tau0,tau2,tau5])
+
+        return -0.5*np.dot(Xall-Tall,np.dot(Xall-Tall,invcov))
+    
+
+    def logProb(self, theta, tau_stats, rowe_stats, ranges, invcov):
+        lp = self.logPrior(theta,ranges)
+        if not np.isfinite(lp):
+            return -np.inf
+        return lp + self.logLike(theta, tau_stats, rowe_stats, invcov)
+
+
+    def compute_alltau(self, gra, gdec, g, gw, s, sra, sdec, e_psf, e_mod, de_psf, T_f):
+        '''
+        Compute tau0, tau2, tau5.
+        All three needs to be computed at once due to covariance.
+
+        gra    : RA of galaxies
+        gdec   : DEC of galaxies
+        g      : shear for observed galaxies np.array((e1, e2)
+        gw     : weights
+
+        s      : indices of stars to use in calculation
+        sra    : RA of stars
+        sdec   : DEC of stars
+        
+        e_psf  : measured ellipticities of PSF from stars -- np.array((e1psf, e2psf))
+        e_mod  : model ellipticities of PSF               -- np.array((e1mod, e2mod))
+        de_psf : e_psf-e_mod                              -- np.array((e1psf, e2psf))
+        T_f    : (T_meas - T_model)/T_meas                -- np.array((e1psf, e2psf))
+        '''
+        
+        p = e_mod
+        q = de_psf
+        w = e_psf * T_f
+        
+        import treecorr
+        sra, sdec = np.array((sra[s], sdec[s])) # Get ra/dec for specific stars
+        p = np.array(( [p[0][s], p[1][s]]))     # Get p for specific stars
+        q = np.array(( [q[0][s], q[1][s]]))     # Get q for specific stars
+        w = np.array(( [w[0][s], w[1][s]]))     # Get w for specific stars
+
+        print(f"Computing Tau 0,2,5 and the covariance")
+        
+        # Load all catalogs
+        catg = treecorr.Catalog(ra=gra, dec=gdec, g1=g[0], g2=g[1], w=gw, ra_units="deg", dec_units="deg",npatch=40) # galaxy shear
+        catp = treecorr.Catalog(ra=sra, dec=sdec, g1=p[0], g2=p[1], ra_units="deg", dec_units="deg",patch_centers=catg.patch_centers) # e_model
+        catq = treecorr.Catalog(ra=sra, dec=sdec, g1=q[0], g2=q[1], ra_units="deg", dec_units="deg",patch_centers=catg.patch_centers) # (e_* - e_model)
+        catw = treecorr.Catalog(ra=sra, dec=sdec, g1=w[0], g2=w[1], ra_units="deg", dec_units="deg",patch_centers=catg.patch_centers) # (e_*(T_* - T_model)/T_* )
+        
+        # Compute all corrleations
+        corr0 = treecorr.GGCorrelation(self.config)
+        corr0.process(catg, catp)
+        corr2 = treecorr.GGCorrelation(self.config)
+        corr2.process(catg, catq)
+        corr5 = treecorr.GGCorrelation(self.config)
+        corr5.process(catg, catw)
+        
+        # Estimate covariance using bootstrap ordering is xip,xim,xip,xim,xip,xim
+        # For our particular purpose we only care about xip so remove xim elements. 
+        cov = treecorr.estimate_multi_cov([corr0,corr2,corr5], 'bootstrap')
+        cov = np.delete(cov,np.arange(100,120),axis=0)
```

Comment:
These numbers won't be correct if the user changes the nbins option.

Reply:
fixed